File: Eikonal_solver/Python/Grid.py
import numpy as np

class Grid:
    """
    Class for sparse storage of numpy_float64 values
    """
    def __init__(self, standard_value, shape):
        self.elements = {}
        self.standard_value = standard_value
        self.shape = shape

    def __setitem__(self, index, value):
        self.elements[index] = value

    def __getitem__(self, index):
        if index in self.elements:
            return self.elements[index]
        return self.standard_value

    def to_array(self):
        array = np.zeros(self.shape)
        array[:] = self.standard_value
        for index in self.elements:
            array[index] = self.elements[index]
        return array

if __name__ == '__main__':
    grid = Grid(0)
    grid[0, 0, 0, 0] = 1
    print grid[0, 0, 0, 0]
    print grid[1, 1, 1, 1]

File: Eikonal_solver/C/Grid.cpp
#include "Grid.h"
#include <fstream>
#include <iostream>
#include <math.h>
using namespace std;

Grid::Grid(double standard_value, int shape[4]) {
    this->standard_value = standard_value;
    this->shape = (struct Numbers) {shape[0], shape[1], shape[2], shape[3]};
}

void Grid::setItem(int array[4], double value) {
    unsigned long long ind = array[3] + (array[2] * 1000)
                          + (array[1] * 1000000) + (array[0] * 1000000000);
    elements[ind] = value;
}

double Grid::getItem(int array[4]) {
    unsigned long long ind = array[3] + (array[2] * 1000) + (array[1] * 1000000) + (array[0] * 1000000000);
    return elements[ind];
}

struct Numbers Grid::getShape() {
    return this->shape;
}

void Grid::writeToFile(char* filename, int n, double xmin, double xmax) {
    ofstream file;
    file.open(filename);
    file << n << " " << xmin << " " << xmax << "\n";
    for (const auto& e : elements) {
        file.fill('0');
        file.width(12);
        file << e.first << " " << e.second << "\n";
    }
    file.close();
}
